xwiki-platform-core/xwiki-platform-notifications/xwiki-platform-notifications-api/src/main/java/org/xwiki/notifications/internal/email/live/LiveMimeMessageIterator.java

```diff
@@ -0,0 +1,137 @@
+
+    @Inject
+    @Named("cached")
+    private ModelBridge modelBridge;
+
+    @Inject
+    private Logger logger;
+
+    /**
+     * Initialize the iterator.
+     *
+     * @param userIterator iterator that returns all users
+     * @param factoryParameters parameters for the email factory
+     * @param event the event that has to be sent to every user
+     * @param templateReference reference to the mail template
+     */
+    public void initialize(NotificationUserIterator userIterator, Map<String, Object> factoryParameters,
+            CompositeEvent event, DocumentReference templateReference)
+    {
+        this.compositeEvent = event;
+        super.initialize(userIterator, factoryParameters, templateReference);
+    }
+
+    /**
+     * For the given user, we will have to check that the composite event that we have to send matches the user
+     * preferences.
+     *
+     * If, for any reason, one of the events of the original composite event is not meant for the user, we clone
+     * the original composite event and remove the incriminated event.
+     */
+    protected List<CompositeEvent> retrieveCompositeEventList(DocumentReference user) throws NotificationException
+    {
+        CompositeEvent resultCompositeEvent = new CompositeEvent(this.compositeEvent);
+
+        if (this.hasCorrespondingNotificationPreference(user, resultCompositeEvent)) {
+            // Apply the filters that the user has defined in its notification preferences
+            // If one of the events present in the composite event does not match a user filter, remove the event
+            for (NotificationFilter filter : this.notificationFilterManager.getAllNotificationFilters(user)) {
+                CopyOnWriteArrayList<Event> events = new CopyOnWriteArrayList<>(resultCompositeEvent.getEvents());
+                for (Event event : events) {
+                    if (filter.filterEvent(event, user, NotificationFormat.EMAIL)) {
+                        resultCompositeEvent.remove(event);
+                        if (resultCompositeEvent.getEvents().size() == 0) {
+                            return Collections.emptyList();
+                        }
+                    }
+                }
+            }
+
+            return Collections.singletonList(resultCompositeEvent);
+        }
+
+        return Collections.emptyList();
+    }
+
+    /**
+     * Test if the given user has enabled the notification preference corresponding to the given composite event.
+     *
+     * @param user the user from which we have to extract
+     * @param compositeEvent
+     * @return
+     */
+    private boolean hasCorrespondingNotificationPreference(DocumentReference user, CompositeEvent compositeEvent)
+    {
+        try {
+            for (NotificationPreference notificationPreference: this.modelBridge.getNotificationsPreferences(user)) {
+                if (notificationPreference.getEventType().equals(compositeEvent.getType())) {
```

Comment:
You need to verify that the preference is about the `email` format, and not the `alert` one (a.ka. the notifications in the menu).